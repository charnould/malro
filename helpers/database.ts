import { Database } from 'bun:sqlite'

export enum DB {
	events = 'events',
	users = 'users'
}

export const mount_db = (db: DB, options?) => new Database(`./datastore/${db}.db`, options)
