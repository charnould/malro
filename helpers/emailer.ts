import { env } from 'bun'

// For sending emails through MailPace (privacy-first emailer)
export const send_email = async (metadata: { subject: string; email: string; to: string }) => {
	if (env.NODE_ENV === 'production') {
		const endpoint = env.EMAILER_ENDPOINT!
		const token = env.EMAILER_TOKEN!

		return await fetch(endpoint, {
			body: JSON.stringify({
				from: 'MALRO.org <robot@malro.org>',
				subject: metadata.subject,
				htmlbody: metadata.email,
				to: metadata.to
			}),
			// prettier-ignore
			headers: {
				'Content-Type': 'application/json',
				'MailPace-Server-Token': token,
				Accept: 'application/json'
			},
			method: 'POST'
		})
	}
	console.log(`In Production, ${metadata.to} receives this email:
    ${metadata.email}`)
}
