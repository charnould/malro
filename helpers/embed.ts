import { format } from 'sql-formatter'
import { env } from 'bun'

import { Options } from '../schema/api'

export const build_options = (data) => {
	const options: Options = {
		display_formatted_query: null,
		referral: data.referral,
		origin: data.origin,
		embed_code: null,
		lang: data.lang,
		html_link: null,
		json_link: null,
		is_valid: null,
		embed: false,
		format: null,
		query: null
	}

	options.query = data.query
	options.embed = data.embed === 'true' ? true : false
	options.format = data.format === 'json' ? 'json' : '001'

	if (typeof options.referral === 'string') {
		const regex = /^<a.+>.+<\/a>$/.test(options.referral)
		options.referral = regex === true ? options.referral : null
	}

	// Prepare/clean SQL query
	let cleaned_query: string

	if (options.origin === 'api') cleaned_query = data.decoded_url.split('/')[4].split('?')[0]
	if (options.origin === 'embed') cleaned_query = options.query
	if (options.origin === 'example') cleaned_query = options.query

	cleaned_query = format(cleaned_query, {
		expressionWidth: 900,
		denseOperators: true,
		keywordCase: 'lower',
		language: 'sqlite',
		tabWidth: 1
	})

	options.query = cleaned_query
		.replace(/\r?\n|\r/g, ' ') // Delete line breaks
		.replace(/\s{2,}/g, ' ') // Delete double spaces
		.replace(/"/g, "'") // Change double quotes for single
	options.html_link = encodeURI(`${env.BASE_URL}/sql/${options.query}`)
	options.json_link = encodeURI(`${env.BASE_URL}/sql/${options.query}?format=json`)
	options.display_formatted_query = format(options.query, {
		indentStyle: 'standard',
		keywordCase: 'upper',
		language: 'sqlite'
	})

	options.embed_code = `<div style="height:460px;width:100%"><iframe width=100% height="100%" src="${options.html_link}?embed=true" frameborder=0 loading=lazy></iframe></div>`

	return options
}
