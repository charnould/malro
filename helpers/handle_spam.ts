// Very rough idea
//
// If a GitHub user is a *Spam Corps* member
// and GitHub Action is `labeled`
// and `label.name` is `spam:yes`,
// then, all events are marked as `spam`
// and will be permanently deleted from DB in 10 dans (with daily DP purge)

// If a GitHub user is a *Spam Corps* member
// and GitHub Action is `labeled`
// and `label.name` is `spam:pending`,
// then, all events are marked as `pending`
// and MALRO user must modify its events to not be considered as spam

export const handle_spam = () => null
