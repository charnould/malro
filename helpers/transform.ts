import { Event } from '../schema/event'
import { LANG } from '../enums/lang'

//
//
//
//
//
//
// Sort all event calendar by start date
// Mainly used in Zod schemas
export const sort_calendar_by_start_date = (array) =>
	array.sort((a: { start: string }, b: { start: string }) => {
		if (a.start < b.start) return -1
		if (a.start > b.start) return 1
		return 0
	})

//
//
//
//
//
//
// Keep only future occurrence of an event
// Useful to display a MALRO card with useful info.
export const keep_only_future_occurrences = (calendar: Event['calendar']) => {
	calendar = calendar.filter((s: (typeof Event)['calendar'], index: number) => {
		s.index = index // Mandatory to get the correct calendar record
		return s.start > new Date().toISOString().substring(0, 16)
	})
	sort_calendar_by_start_date(calendar)
	return calendar
}

//
//
//
//
//
//
// Get relevant translation of an event description in regards of user preferred langs
// Useful to display a MALRO card with useful info.
// TODO: test
export const get_relevant_translation = (description: Event['description'], langs: typeof LANG) => {
	let relevant_description: (typeof Event)['description']
	for (const lang of langs) {
		relevant_description = description.find((d) => d.lang === lang)
		if (relevant_description !== undefined) break
	}
	return [relevant_description]
}

//
//
//
//
//
//
// Transform a date into ICAL format
// TODO: test
export const transform_date = (date?: undefined | string) => {
	if (date === undefined) {
		return `${new Date().toISOString().replace(/[-:.]/g, '').substring(0, 13)}00`
	}
	return `${date.replace(/[-:]/g, '')}00`
}

//
//
//
//
//
//
// Parse a DB query (0, 1 or more events) into a JS-event/object usable in code
//
export const turn_stringified_into_parsed_events = (events: any) => {
	for (const event of events) {
		if (event !== null) {
			event.mandatory_booking = JSON.parse(event.mandatory_booking)
			event.booking_medium = JSON.parse(event.booking_medium)
			event.description = JSON.parse(event.description)
			event.linked_to = JSON.parse(event.linked_to)
			event.featuring = JSON.parse(event.featuring)
			event.calendar = JSON.parse(event.calendar)
			event.type = JSON.parse(event.type)
		}
	}

	return events as Event[]
}

//
//
//
//
//
//
// Transform firstname `piotr illich charles-henri` into `Piotr Illich Charles-Henri`
export const capitalize_firstname = (firstname: string) => {
	const firstnames = firstname.trim().replace(/\s\s+/g, ' ').toLowerCase().split(' ')
	for (let i = 0; i < firstnames.length; i++) {
		firstnames[i] = firstnames[i].charAt(0).toUpperCase() + firstnames[i].slice(1)
		if (firstnames[i].includes('-')) {
			const hyphenated_name = firstnames[i].split('-')
			for (let i = 0; i < hyphenated_name.length; i++) {
				hyphenated_name[i] = hyphenated_name[i].charAt(0).toUpperCase() + hyphenated_name[i].slice(1)
			}
			firstnames[i] = hyphenated_name.join('-')
		}
	}
	return firstnames.join(' ')
}

//
//
//
//
//
//
// Strip Emojis
export const strip_multiple_spaces_and_emojis = (string: string) =>
	string
		.replace(
			/([\u2700-\u27BF]|[\uE000-\uF8FF]|\uD83C[\uDC00-\uDFFF]|\uD83D[\uDC00-\uDFFF]|[\u2011-\u26FF]|\uD83E[\uDD10-\uDDFF])/g,
			''
		)
		.replace(/\s+/g, ' ')
		.trim()
