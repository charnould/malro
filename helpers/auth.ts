import { getCookie } from 'hono/cookie'
import type { Context, Next } from 'hono'
import * as jose from 'jose'
import { env } from 'bun'
import { z } from 'zod'

import { click_magic_link, ask_admin } from '../views/emails/login'
import { get_users, get_user } from '../models/users'
import { get_event, Table } from '../models/events'
import { mount_db, DB } from './database'
import { send_email } from './emailer'
import type { User } from '../schema/user'

const db = mount_db(DB.users)

//
//
//
//
//
//
export const extract_domain_from_email = (email: User['email']) =>
	email.toLowerCase().trim().split('@')[1].split('.').slice(-2).join('.')

//
//
//
//
//
//
// is_user_authorized()
// Check if a user trying to login/sign up is allowed to do so.
// Two cases are considered: very first user and next users.
//

// Types
export const Is_user_authorized = z.enum([
	'domain_banned',
	'new_organization',
	'user_already_exists',
	'user_need_admin_permission'
])

// eslint-disable-next-line no-redeclare
export type Is_user_authorized = z.infer<typeof Is_user_authorized>

/* Function */
export const is_user_authorized = async (email: string) => {
	const domain = extract_domain_from_email(email)

	/**
	 * malro bans free email provider domains (like Gmail, iCloud, etc.) to prevent spamming.
	 * Thus, if Domain is banned, User cannot use malro.
	 */
	let is_domain_authorized = true
	const query = await db.prepare('SELECT * FROM banned WHERE domain = ?').get(domain)
	if (query !== null) is_domain_authorized = false

	if (!is_domain_authorized) return 'banned_domain'

	/**
	 * If Domain is allowed and it's the very first user for a domain trying to sign up/login,
	 * he/she is authorized and is granted administrator role. Admin is then the only
	 * one able to grant access to other users for this domain through its `settings`.
	 */
	const users = get_users(domain)
	if (users.length === 0) return 'new_organization'

	/**
	 * If Domain is allowed and Email is in Users DB
	 * that means that User is allowed to use malro.
	 */
	if (users.some((user) => user.email === email)) return 'user_already_exists'

	/**
	 * Consequently, in all other cases,
	 * that means that User need admin permission.
	 */
	return 'user_needs_admin_permission'
}

//
//
//
//
//
//
// Send a magic link by email to a user
export const send_magic_link_to = async (email: string) => {
	const token = await encrypt_jwt({ email: email }, env.CRYPTO_1, '1d')
	const magic_link = `${env.BASE_URL}/token?verify=${token}`
	const message = click_magic_link(magic_link).toString()
	await send_email({ subject: 'Login to MALRO.org', email: message, to: email })
	return
}

//
//
//
//
//
//
// Send a list of admin emails to a user that wants to be granted an access
export const send_admin_emails_to = async (email: string) => {
	const domain = extract_domain_from_email(email)
	const users = get_users(domain)
	const admins = users
		.filter((user: User) => user.is_admin === 'true')
		.map((user: User) => user.email)
		.sort()
	const message = ask_admin(admins).toString()
	await send_email({ subject: 'Login to MALRO.org', email: message, to: email })
	return
}

//
//
//
//
//
//
// is_auth() Middleware
// Check for each request if the user is authenticated or not
// and respond accordingly: move on or redirect
export const is_auth = async (c: Context, next: Next) => {
	try {
		// Decypher email
		const cookie = getCookie(c, 'malro')
		const { email }: User['email'] = await decrypt_jwt(cookie, env.CRYPTO_2)

		// Case 1. Check if user exists in DB
		const user = await get_user(email)
		if (user === undefined) return c.redirect('/')

		// Case 2. Check if user has access to this event/draft
		const kind = c.req.param('kind')
		if (kind === 'event' || kind === 'draft') {
			let created_by
			const id = c.req.param('id')
			if (kind === 'draft') created_by = get_event(id, Table.drafts)?.created_by
			if (kind === 'event') created_by = get_event(id, Table.events)?.created_by

			const domain = extract_domain_from_email(email)
			if (domain !== created_by) return c.text("ðŸ‘€ â€” You don't have access to this event.")
		}

		// Otherwise, user can access this route!
		c.set('user', user)
		return next()
	} catch {
		return c.redirect('/')
	}
}

//
//
//
//
//
//
// is_setup() Middleware
// Check if user has set up its account (lang)
export const is_setup = async (c: Context, next: Next) => {
	if (c.get('user').lang === '') return c.redirect('/settings?setup=true')
	return next()
}

//
//
//
//
//
//
// Encrypt JWT
// Types
const Payload = z.object({ email: z.string().email() })
type Payload = z.infer<typeof Payload>

export const encrypt_jwt = async (payload: Payload, secret: string, expiration_time: string) => {
	const key = jose.base64url.decode(secret)
	return await new jose.EncryptJWT(payload)
		.setProtectedHeader({ enc: 'A128CBC-HS256', alg: 'dir' })
		.setIssuer('malro_server')
		.setAudience('malro_user')
		.setExpirationTime(expiration_time) // ex. "1h"
		.setIssuedAt()
		.encrypt(key)
}

//
//
//
//
//
//
// Decrypt JWT
export const decrypt_jwt = async (token: string, secret: string) => {
	const key = jose.base64url.decode(secret)
	const { payload } = await jose.jwtDecrypt(token, key, {
		issuer: 'malro_server',
		audience: 'malro_user'
	})
	return payload
}
