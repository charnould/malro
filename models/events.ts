// ALL TESTED

import type { Statement } from 'bun:sqlite'
import { spawnSync, argv } from 'bun'

import { turn_stringified_into_parsed_events } from '../helpers/transform'
import { mount_db, DB } from '../helpers/database'
import type { Event } from '../schema/event'
import type { User } from '../schema/user'

const events_db = mount_db(DB.events)
const users_db = mount_db(DB.users)

export enum Table {
	events = 'events',
	drafts = 'drafts'
}

//
//
//
//
//
//
export const get_event = (id: Event['id'], table: Table) => {
	if (table === Table.events) return events_db.prepare('SELECT * FROM events WHERE id = ?').get(id)
	if (table === Table.drafts) return users_db.prepare('SELECT * FROM drafts WHERE id = ?').get(id)
}

//
//
//
//
//
//
export const save_event = (event: Event, table: Table) => {
	let stmt: Statement

	if (table === Table.events) {
		stmt = events_db.prepare(`
    INSERT OR REPLACE INTO events (id, created_by, status, updated_at, linked_to, longitude, latitude, country, zipcode, city, street, timezone, type, image_url, video_url, min_age, max_age, door_time, mandatory_booking, booking_start, booking_medium, currency, description, calendar, support_email, support_phone, featuring)
    VALUES                        (? , ?         , ?     , ?         , ?        , ?        , ?       , ?      , ?      , ?   , ?     , ?       , ?    , ?         , ?          , ?      , ?      , ?        , ?                , ?            , ?             , ?       , ?          , ?       , ?            , ?            , ?        )`)

		users_db.prepare('DELETE FROM drafts WHERE id = ?').run(event.id)
	}

	if (table === Table.drafts) {
		stmt = users_db.prepare(`
    INSERT OR REPLACE INTO drafts (id, created_by, status, updated_at, linked_to, longitude, latitude, country, zipcode, city, street, timezone, type, image_url, video_url, min_age, max_age, door_time, mandatory_booking, booking_start, booking_medium, currency, description, calendar, support_email, support_phone, featuring)
    VALUES                        (? , ?         , ?     , ?         , ?        , ?        , ?       , ?      , ?      , ?   , ?     , ?       , ?    , ?         , ?          , ?      , ?      , ?        , ?                , ?            , ?             , ?       , ?          , ?       , ?            , ?            , ?        )`)
	}

	stmt.run(
		event.id,
		event.created_by,
		event.status,
		new Date().toISOString(),
		JSON.stringify(event.linked_to),
		event.longitude,
		event.latitude,
		event.country,
		event.zipcode,
		event.city,
		event.street,
		event.timezone,
		JSON.stringify(event.type),
		event.image_url,
		event.video_url,
		event.min_age,
		event.max_age,
		event.door_time,
		JSON.stringify(event.mandatory_booking),
		event.booking_start,
		JSON.stringify(event.booking_medium),
		event.currency,
		JSON.stringify(event.description),
		JSON.stringify(event.calendar),
		event.support_email,
		event.support_phone,
		JSON.stringify(event.featuring)
	)

	return
}

//
//
//
//
//
//
export const list_events = (domain: User['domain']) => {
	// Get published events from events.db
	const events = events_db.prepare('SELECT * FROM linked_to INNER JOIN events USING(id) WHERE domain= ?').all(domain)
	const drafts = users_db.prepare('SELECT * FROM drafts WHERE created_by = ?').all(domain)

	const all = [...events, ...drafts]
		.sort((a: Event['updated_at'], b: Event['updated_at']) => a.updated_at - b.updated_at)
		.reverse()

	return turn_stringified_into_parsed_events(all)
}

//
//
//
//
//
//
export const trash_organization = (domain: User['domain']) => {
	// First, remove any mention of `domain` in `linked_to`
	// (indeed, if an org is trashed, it must not be linked to any event)...
	let events = events_db.query('SELECT * FROM events WHERE linked_to LIKE ?').all(`%"${domain}"%`)
	events = turn_stringified_into_parsed_events(events)
	for (const event of events) {
		event.linked_to = event.linked_to.filter((d: string) => d !== domain)
		save_event(event, Table.events)
	}

	// ...trash events created by `domain`...
	events_db.query(`UPDATE events SET status = "trashed" WHERE created_by = ?`).run(domain)

	// ...and finally delete all users and drafts
	users_db.query('DELETE FROM drafts WHERE created_by = ?').run(domain)
	users_db.query('DELETE FROM users WHERE domain = ?').run(domain)

	return
}

//
//
//
//
//
// A hack to be able to trash_organization() using bash.
// Tightly link to `bash run.sh trash_organization`.
// Useful when someone has made some tests in production, is not a spam
// and you just want to keep MALRO DB clean.
if (argv[2] === 'trash_organization') {
	const domain = argv[3].trim().toLowerCase()
	trash_organization(domain)
}

//
//
//
//
//
//
export const archive_organization = (domain: User['domain']) => {
	// Delete all unneeded stuff...
	users_db.query('DELETE FROM drafts WHERE created_by = ?').run(domain)
	users_db.query('DELETE FROM users WHERE domain = ?').run(domain)

	// ...and, for each event,
	// check if we keep it for posterity
	// or if we can trash it safely
	let events = events_db.query('SELECT * FROM events WHERE created_by = ?').all(domain)
	events = turn_stringified_into_parsed_events(events)

	for (const event of events) {
		event.calendar = event.calendar.filter(
			(s: { start: string }) => s.start < new Date().toISOString().substring(0, 16)
		)

		// if `event.calendar` IS empty,
		// it means that this event did not occured in the past.
		// Hence, no need to keep this event in DB for posterity.
		// FYI: Promotional image will be removed with vaccuum cronjob
		if (event.calendar.length === 0) trash_event(event.id)
		// if `event.calendar` IS NOT empty,
		// it means that we should keep this already-occured event for posterity
		else {
			save_event(event, Table.events)
		}
	}

	return
}

//
//
//
//
//
//
export const link_an_organization_to_an_event = (id: Event['id'], domain: User['domain']) => {
	const event = get_event(id, Table.events)
	if (event !== null) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
		const e: Event = turn_stringified_into_parsed_events([event])[0]
		e.linked_to = [...e.linked_to, domain].sort()
		e.linked_to = [...new Set(e.linked_to)]
		save_event(e, Table.events)
		return 'ORGANIZATION_LINKED'
	}
	return 'NO_EVENT_WITH_THIS_UUID'
}

//
//
//
//
//
//
export const unlink_an_organization_from_an_event = (id: Event['id'], domain: User['domain']) => {
	const event = get_event(id, Table.events)
	if (event !== null) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
		const e: Event = turn_stringified_into_parsed_events([event])[0]
		e.linked_to = e.linked_to.filter((d: string) => d !== domain)
		save_event(event, Table.events)
		return 'ORGANIZATION_UNLINKED'
	}
	return 'NO_EVENT_WITH_THIS_UUID'
}

//
//
//
//
//
//
export const duplicate_event = (id: Event['id'], domain: User['domain']) => {
	let event = get_event(id, Table.events)
	event = turn_stringified_into_parsed_events([event])[0]
	event.updated_at = Math.round(Date.now() / 1000)

	for (const d of event.description) {
		d.title = `[COPY] ${d.title}`
	}
	event.linked_to = `["${domain}"]`
	event.id = crypto.randomUUID()
	event.status = 'drafted'
	save_event(event, Table.drafts)
	return
}

//
//
//
//
//
//
export const trash_event = (id: Event['id']) => {
	users_db.query('DELETE FROM drafts WHERE id = ?').run(id)
	// TODO: handle image where you delete a draft

	let event = events_db.query('SELECT * FROM events WHERE id = ?').all(id)
	event = turn_stringified_into_parsed_events(event)[0]
	event.status = 'trashed'
	event.linked_to = []
	event.calendar = []
	save_event(event, Table.events)
	return
}

//
//
//
//
//
//
export const vacuum_trashed_events = () => {
	const db = mount_db(DB.events)
	// Get a date 10 days in the past,
	// then delete all old trash events

	const d1 = new Date()
	const d2 = d1.setDate(d1.getDate() - 10)
	const ten_days_ago = new Date(d2).toISOString()

	const images_to_delete: string[] = db
		.prepare('SELECT id FROM events WHERE updated_at < ? AND status = "trashed"')
		.all(ten_days_ago)

	for (const image of images_to_delete) {
		spawnSync(['rm', `datastore/images/${image.id}.webp`])
		spawnSync(['rm', `datastore/images/originals/${image.id}.webp`])
	}

	db.prepare('DELETE FROM events WHERE updated_at < ? AND status = "trashed"').run(ten_days_ago)

	console.info('[ðŸ¤– â€” cronbot]: DB and datastore/images vacuumed')
	return
}

//
//
//
//
//
//
export const count_events = () => {
	const db = mount_db(DB.events)
	const published = db.prepare('SELECT COUNT(*) as count FROM events WHERE status = "published"').get()
	const trashed = db.prepare('SELECT COUNT(*) as count FROM events WHERE status = "trashed"').get()
	return { published: published.count, trashed: trashed.count }
}
