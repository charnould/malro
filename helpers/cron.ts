import { Cron } from 'croner'
import { spawn } from 'bun'

import { update_third_party_banned_domains } from '../models/users'
import { vacuum_trashed_events } from '../models/events'
import { send_report } from './send_report'

export const launch_cronjobs = () => {
	// Update banned domains using a third party list
	// Runs every monday at 01:00
	// Test: ok
	// TODO: mettre la bonne date aprÃ¨s import initial
	Cron('10 9 * * 3', () => update_third_party_banned_domains())

	// Delete old trashed events and spams from events DB
	// Runs every day at 01:00
	// TODO: vacuum DB (sqlite pragma)
	// Test: NOK
	Cron('0 1 * * *', () => vacuum_trashed_events())

	// Archive and gzip events db
	// Runs every day at 23:45
	// Test: ok
	Cron('45 23 * * *', () => spawn(['bash', 'run.sh', 'archive_db']))

	// Send a "Daily Report" to charnould@icloud.com
	// Runs every day at 23:59
	Cron('59 23 * * *', () => send_report())

	// TODO: (?) Delete old archives
	// TODO: (?) Identify duplicates and inform users
	// TODO: (?) inform user to update a perpetual event with new dates
	return
}
