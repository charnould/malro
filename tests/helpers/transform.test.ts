import { expect, it } from 'bun:test'

import {
  strip_multiple_spaces_and_emojis,
  keep_only_future_occurrences,
  sort_calendar_by_start_date,
  capitalize_firstname
} from '../../helpers/transform'
import { turn_stringified_into_parsed_events } from '../../helpers/transform'
import { save_event, get_event, Table } from '../../models/events'
import { dummy_event } from '../dummy/events'

//
//
//
//
//
//
it('should sort calendar by start date', () => {
  const input = [
    { start: '2023-01-02T12:00', end: '2023-01-02T21:00' },
    { start: '2001-01-01T12:00', end: '2001-01-01T21:00' },
    { start: '2022-01-01T12:00', end: '2022-01-01T21:00' }
  ]
  expect(sort_calendar_by_start_date(input)).toMatchSnapshot()
})

//
//
//
//
//
//
it('should keep only future occurrences of an event', () => {
  save_event(dummy_event, Table.drafts)
  const stringified_event = get_event(dummy_event.id, Table.drafts)

  const event = turn_stringified_into_parsed_events([stringified_event])[0]
  const only_future_occurrences = keep_only_future_occurrences(event.calendar)
  expect(only_future_occurrences).toMatchSnapshot()
})

//
//
//
//
//
//
it('should capitalize firstname', () => {
  const input = ' piotr ILlich d. j michel-ange jean-sébastien  '
  const output = capitalize_firstname(input)
  expect(output).toBe('Piotr Illich D. J Michel-Ange Jean-Sébastien')
})

//
//
//
//
//
//
it('should strip multiple spaces and emojis', () => {
  const input = ' 👍 This 👍👍 is a     👍 👍  👍 test 👍  '
  const output = strip_multiple_spaces_and_emojis(input)
  expect(output).toBe('This is a test')
})
