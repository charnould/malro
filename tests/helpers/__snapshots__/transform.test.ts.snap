// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should sort calendar by start date 1`] = `
[
  {
    "end": "2001-01-01T21:00",
    "start": "2001-01-01T12:00",
  },
  {
    "end": "2022-01-01T21:00",
    "start": "2022-01-01T12:00",
  },
  {
    "end": "2023-01-02T21:00",
    "start": "2023-01-02T12:00",
  },
]
`;

exports[`should keep only future occurrences of an event 1`] = `
[
  {
    "adult_fee": 15,
    "audio_lang": "fr",
    "child_fee": 5,
    "end": "2030-10-01T21:00",
    "feature": null,
    "index": 1,
    "start": "2030-10-01T19:30",
  },
  {
    "adult_fee": 35,
    "audio_lang": "fr",
    "child_fee": 10,
    "end": "2100-06-12T14:30",
    "feature": null,
    "index": 2,
    "start": "2100-06-12T12:30",
  },
]
`;
